Phylo csf++ bed included

pskvins opened the pull request:
Hello Doctor Christopher Pockrandt,
I'm who asked about issue that happens with Drosophila melanogaster chromosome 2L([https://github.com/cpockrandt/PhyloCSFpp/issues/5](url))

I'm asking pull request of code, which is for generating bed file that gives the information about coding region.
It makes bed file when --output-phylo is true in build-tracks.

I used the same command as I showed you in the issue, but I used the whole genome length as you told me to.

I compared the result file with the file 'PhyloCSF+2.bed.gz' that I added below.
[PhyloCSF+2.bed.gz](https://github.com/cpockrandt/PhyloCSFpp/files/6323736/PhyloCSF%2B2.bed.gz)
As a result, among 13334 regions in PhyloCSF+2.bed.gz, 13108 regions are perfectly matched, 37 regions are found only in PhyloCSF+2.bed.gz, 202 regions are found only in result file, and 325 regions in PhyloCSF+2.bed.gz are more than 95% matched (but not perfectly matched) with the result file.

Please take a look on the code that I added, and let me know if I have to refine any of my code.


Thank you.

cpockrandt commented:
Thank you for your work and the pull-request @pskvins 

I'm still testing your feature and will come back when it finished. There are a few `TODO`s in your code. Are they still relevant or can they be deleted? 

I have two major suggestions:

1. can you add a new flag `--output-regions BOOL` and turn it off by default (like `--output-phylo BOOL`) so users can turn it on and off as they like?

2. you deleted a test file chr22_25_28_each_30k_bases.maf.gz as well as added some whitespace changes that were probably introduced by your editor. Can you undo that so that the PR is cleaner and doesn't change anything that does not affect your code? Then the tests on travis should also pass.

Christopher

pskvins commented:
Thank you for reviewing my code, Christopher.

I added --output-regions BOOL and make it to show the error message when people try to run --output-regions when phylo_smooth is false.
I checked if the added function work well in three different conditions: 
where I omitted --output-regions,
where I added --output-regions 0,
and for --output-regions 1.

I also added the explanation about --output-regions and checked if --help shows it well, 
recovered a test file chr22_25_28_each_30k+bases.maf.gz,
 and undid all the whitespace that I could find.

Please take a look on the modification and let me know about the things I have to correct more.

Thank you.

Sukhwan

cpockrandt reviewed (changes requested):
Thank you, it looks a lot cleaner now! :) I added a few comments to your code (you should be able to see them in between the code lines and can comment or mark them as fixed when you changed it).

Please have a look at: https://github.com/cpockrandt/PhyloCSFpp/pull/6/checks
When you open a PR or make changes, the code will be compiled and tested. A few compilers fail because they believe that you have an uninitialized value (starting_count). Your code might guarantee that the value will be initialized before reading it, but the compiler fails to see this. We want to avoid any errors popping up when users compile the software themselves.

We have a very small test example that we also use for testing [1]. Can you add ``--output-regions`` there and also store the expected result file in ./test/expected_results/build-tracks (after making sure that the output is indeed correct)?

I haven't looked at the actual results and the differences to the PhyloCSF regions. Did you investigate the differences? With PhyloCSF++ we basically guarantee that users get the very same results as with the original PhyloCSF software, so we need to be very certain that any differences are really due to randomization and that there are no bugs. It seems your code does not do any randomization, so if you get different regions, that should only happen if we also get different smoothed PhyloCSF scores. If I remember correctly, the differences in smoothed PhyloCSF scores are very marginal though, and I'm not sure whether this could explain that many differences.

As a last note: it might help to rebase your branch (``git fetch --all && git rebase master`` and a (forced) push afterwards), so that the PR does only contain your changes and no changes from me that I made on the master branch in the meantime.

pskvins commented on src/create_tracks.hpp:
```diff
@@ -225,14 +238,69 @@ double compute_log_odds(double prob) {
         double log_odds = compute_log_odds(state_probabilities[cur_codon_count][0]);
         result.emplace_back(chunk_start_pos, chunk_end_pos, log_odds);
     }
+    //TODO fill bedresult with start and end positon of 0 blocks
```
I erased this TODO now.

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -11,6 +11,7 @@
 #include "run.hpp"
 
 #include <sstream>
+#include <iostream>
```
Done removing

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -217,15 +229,41 @@ void run_tracks(const std::string & alignment_path, const Model & model, const T
                                     my_fprintf(file_score[file_index], "%.3f", region[i].log_odds_prob);
                                 }
-
+                                //TODO write be output code
```
I erased this TODO now.

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -242,15 +280,42 @@ void run_tracks(const std::string & alignment_path, const Model & model, const T
                             my_fprintf(file_score[file_index], "%.3f", region[i].log_odds_prob);
                         }
-
+                        //TODO write be output code
```
I erased this TODO now.

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -361,6 +442,12 @@ int main_build_tracks(int argc, char **argv)
     }
 
+    // this has to hold true: output-regions => phylo_smooth
+    if (args.is_set("output-regions") && !(params.phylo_smooth))
```
Changed to params.phylo_bed_file

pskvins commented on src/phylocsf++annotate_with_tracks.hpp:
```diff
@@ -302,4 +302,4 @@ int main_annotate_with_tracks(int argc, char** argv)
 
     return 0;
-}
+}
```
Added newline now

pskvins commented on src/parallel_file_reader.hpp:
```diff
@@ -685,4 +626,4 @@ class parallel_maf_reader
         free(bytes_processing);
     }
-};
+};
```
Added newline now

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -217,15 +229,41 @@ void run_tracks(const std::string & alignment_path, const Model & model, const T
+                                process_scores(model._hmm, scores, startBlockPos, region, bedregions);
 
-                                for(size_t i = 0; i < region.size(); i++)
+                                for (size_t i = 0; i < region.size(); i++)
```
I changed code of there a little bit. I didn't remove the whitespace yet, but do you want me to remove the whitespace there still?

pskvins commented on src/phylocsf++build_tracks.hpp:
```diff
@@ -323,6 +401,7 @@ int main_build_tracks(int argc, char **argv)
     args.add_option("power-threshold", ArgParse::Type::FLOAT, "Minimum confidence score to output PhyloCSF score. Default: " + std::string(threshold_default_str), ArgParse::Level::GENERAL, false);
     args.add_option("genome-length", ArgParse::Type::INT, "Total genome length (needed for --output-phylo).", ArgParse::Level::GENERAL, false);
     args.add_option("coding-exons", ArgParse::Type::STRING, "BED-like file (chrom name, strand, phase, start, end) with coordinates of coding exons (needed for --output-phylo).", ArgParse::Level::GENERAL, false);
+    args.add_option("output-regions", ArgParse::Type::BOOL, "Generate bed files with coordinates of potential protein coding regions (need to put --output-phylo as 1 to run the command). Default: " + std::to_string(params.phylo_bed_file), ArgParse::Level::GENERAL,false);
```
I put --output-raw-phylo, --output-phylo, --output-regions together.
The name of the flag was --output-raw-phylo, do you want me to change --output-raw-phylo to --output-phylo-raw?

pskvins commented:
I made --output-regions to work without --output-phylo being true, and made changes you requested except putting result in ./test/expected_results/build-tracks because I wanted you to check if the result bed file looks valid.

I did a same test using the chr2L of drosophila melanogaster as I told you at the beginning([https://github.com/cpockrandt/PhyloCSFpp/issues/5](url)).
As a result, I got PhyloCSF+2_C++.bed, and compared with PhyloCSF+2_chr2L.bed file which I only captured informations of chr2L from original bed file.

[PhyloCSF+2_chr2L.bed.gz](https://github.com/cpockrandt/PhyloCSFpp/files/6348742/PhyloCSF%2B2_chr2L.bed.gz)

[PhyloCSF+2_C++.bed.gz](https://github.com/cpockrandt/PhyloCSFpp/files/6348741/PhyloCSF%2B2_C%2B%2B.bed.gz)

Among 13334 potential CDSs from original file, 13108 CDSs (98.3%) are found to be exactly same with the resulted bed file, and 37 CDSs (0.28%) are found only (without any overlap) in original file. Resulted bed file has 13690 potential CDSs, which has 202 CDSs that are found only in resulted file, and 379 CDSs are found to be the fragments (overlap 100%, but not exactly same) of CDSs from original file.

I attached a table below that shows what I've explained above. 

[comparing original and generated PhyloCSF2+.bed.pptx](https://github.com/cpockrandt/PhyloCSFpp/files/6348890/comparing.original.and.generated.PhyloCSF2%2B.bed.pptx)

Please take a look and let me know if I can add result bed files to expected_results/build-tracks

pskvins commented:
I've compared PhyloCSF wig files and raw wig files, and find out that while raw scores are almost exactly same in every place, whereas scores in wig files are different largely only in -1 and -3 frames, different moderately in -2 frame, and almost same in every place for positive strands.

I find out that differences in the scores are mostly generated from the codon only appears in the either side of wig files, and also from codons that are near the codon only appears in the either side of wig files, and thought that gap would have caused forward-backward algorithm to compute the score near the gap different from the original file. 

I counted all the gaps in both wig files, and it turns out that for positive strands, there are 2 or no gaps between two files, whereas -1 and -3 frames have about 12,000 gaps on each file and -2 frame has about 230 gaps on each file out of about 7 million codons in each wig file.

I've counted the number of potential coding regions from bed files that are not exactly same with each other, and for positive strands and -2 frame, about 200 are counted from original bed file and about 520 regions are counted from the file I generated. But in -1 and -3 frames, about 415 regions are counted from original file and 769 and 804 regions are counted from -1 and -3 frame respectively. In here, the difference between two cases might have been caused by the score difference in frame -1 and -3.
When I compared the scores of wig files, there was about 1480 codons that the signs of scores are different in positive strand and frame -2, when -1 and -3 frames have about 2330 codons of different signs of scores. Also, among this codons in positive strand and frame -2, score that I generated tend to show more high scores (about 60%), when frame -1 and -3 showed equal percentage.

Lastly, while I run a program, I found one issue about frame. I generated datas from dm6 msa, and I found the negative frame number is off by 1 from the negative frame of original wig file (for example, frame -1 of wig file I generated was same as frame -2 of original PhyloCSF data)

Please take a look and let me know if you need more information or if I need to correct my code.

Thank you.
Sukhwan Park

cpockrandt commented:
Thank you for looking so much into it! I discussed it with Martin and we agree that your code looks good and the discrepancies are most likely due to the smoothing. I will look into it.

I have just one tiny change before we merge it: Can you name the bed files the same way as the official PhyloCSF bed files? i.e., "PhyloCSF+1Regions.bed" instead of "PhyloCSF+1.bed". Thank you! :)

cpockrandt commented on src/create_tracks.hpp:
```diff
@@ -257,9 +257,37 @@ void process_scores(hmm const & hmm, std::vector<double> &scores,
+            if (path[i+1]!=0) {
+                prob = state_probabilities[starting_count][0];
+                double log_odd = compute_log_odds(prob);
+                if (log_odd < 1) {
```
This code fragment appears multiple times now, so it would make the code more readable and easier to maintain by moving it into a separate function like

```
uint32_t compute_bed_color(const double log_odd)
{
    if (log_odd < 1)
        return 210;
    else if (log_odd > 7)
        return 0;
    else
        return 210 - 30 * int(floor(log_odd));
}
```

pskvins commented:
I changed the name of the bed files and made the function of calculating color code to get rid of the redundancies.

I also found the cause of the track scores from negative frames. It was because of the starting position of bls_pos from `for (size_t xx = 0, bls_pos = aln.skip_bases; xx < lpr_per_codon[thread_id].size(); ++xx, bls_pos += 3)`
For negative strand, it seems like it has to take bls_pos = aln.length() % 3 rather than aln.skip_bases, so I splitted up the code for the case of positive strand and negative strand.
I've already tried to compare the resulted track files with the original track files, and almost every score look same. I attached a graph below for comparing score of PhyloCSF and PhyloCSF++ after I changed a score.
![compared_new_wig-3](https://user-images.githubusercontent.com/80232253/118622860-a86b9400-b802-11eb-823c-54e7ce3a04ea.png)
![compared_new_wig-2](https://user-images.githubusercontent.com/80232253/118622865-aa355780-b802-11eb-83f7-0b5e3851371c.png)
![compared_new_wig-1](https://user-images.githubusercontent.com/80232253/118622870-aa355780-b802-11eb-958f-6f61519d453d.png)
![compared_new_wig+3](https://user-images.githubusercontent.com/80232253/118622871-aacdee00-b802-11eb-91e3-7016b9907602.png)
![compared_new_wig+2](https://user-images.githubusercontent.com/80232253/118622876-ab668480-b802-11eb-88ae-adda3d5439f7.png)
![compared_new_wig+1](https://user-images.githubusercontent.com/80232253/118622879-abff1b00-b802-11eb-8372-27556a9f0682.png)

Lastly, there was also a difference between PhyloCSF and PhyloCSF++ about frames in negative strand. It seems like the way of deciding frame on the negative strand is different. Due to this, it seems like it is little bit complex to compare scores from PhyloCSF++ with scores from PhyloCSF, especially when the file contains the informations about multiple chromosomes. I heard that you already know about this, but I wanted to tell about it just in case.

If there is anything to change, please tell me.
Thank you!

cpockrandt commented:
It seems travis did not run for your latest commit, so I will close and reopen your PR, which will hopefully trigger a build.

cpockrandt commented:
Thank you @pskvins, I did the following two things:

- I rebased your commits on master (there was a conflict because of changes on master in the meantime, and your PR is going to go onto the master branch).
- I made some changes in the commit "Code improvements"

This includes:
- removal of some code duplication
- do not write bed files when the user did not set the flag (instead of deleting them afterwards)
- I made a lot of minor changes to reduce the "diff" to master, so it is easier to understand the changes in this commit. When I merge the commit, I will squash all the commits together
- and some other changes that I believe make the code easier to read or maintain

Before merging, I will check the results myself (might take 1-2 days), because I noticed that also the -2 strand and the raw files changed, and I want to make sure that I did not mess up your code with my changes. I'm happy that you found and fixed the bug!

cpockrandt commented:
LGTM!
